pkg/controller/utils/datadog/datadog.go

```diff
@@ -0,0 +1,441 @@
+func (dd *MetricsForwarder) updateCredsIfNeeded(apiKey, appKey string) error {
+	if dd.keysHash != hashKeys(apiKey, appKey) {
+		return dd.initAPIClient(apiKey, appKey)
+	}
+	return nil
+}
+
+// validateCreds returns validates the creds by querying the Datadog API
+func (dd *MetricsForwarder) validateCreds(apiKey, appKey string) (*api.Client, error) {
+	return dd.delegator.delegatedValidateCreds(apiKey, appKey)
+}
+
+// delegatedValidateCreds is separated from validateCreds to facilitate mocking the Datadog API
+func (dd *MetricsForwarder) delegatedValidateCreds(apiKey, appKey string) (*api.Client, error) {
+	datadogClient := api.NewClient(apiKey, appKey)
+	valid, err := datadogClient.Validate()
+	if !valid || err != nil {
+		return nil, fmt.Errorf("datadog client is invalid: %v", err)
+	}
+	return datadogClient, nil
+}
+
+// sendStatusMetrics forwards metrics for each component deployment (agent, clusteragent, clustercheck runner)
+// based on the status of DatadogAgentDeployment
+func (dd *MetricsForwarder) sendStatusMetrics(status *datadoghqv1alpha1.DatadogAgentDeploymentStatus) error {
+	if status == nil {
+		return errors.New("nil status")
+	}
+	var metricValue float64
+
+	// Agent deployment metrics
+	if status.Agent != nil {
+		if status.Agent.Available == status.Agent.Desired {
+			metricValue = deploymentSuccessValue
+		} else {
+			metricValue = deploymentFailureValue
+		}
+		tags := append(dd.tags, fmt.Sprintf(stateTagFormat, string(status.Agent.State)))
+		if err := dd.sendDeploymentMetric(metricValue, agentName, tags); err != nil {
+			return err
+		}
+	}
+
+	// Cluster Agent deployment metrics
+	if status.ClusterAgent != nil {
+		if status.ClusterAgent.AvailableReplicas == status.ClusterAgent.Replicas {
+			metricValue = deploymentSuccessValue
+		} else {
+			metricValue = deploymentFailureValue
+		}
+		tags := append(dd.tags, fmt.Sprintf(stateTagFormat, string(status.ClusterAgent.State)))
+		if err := dd.sendDeploymentMetric(metricValue, clusteragentName, tags); err != nil {
+			return err
+		}
+	}
+
+	// Cluster Check Runner deployment metrics
+	if status.ClusterChecksRunner != nil {
+		if status.ClusterChecksRunner.AvailableReplicas == status.ClusterChecksRunner.Replicas {
+			metricValue = deploymentSuccessValue
+		} else {
+			metricValue = deploymentFailureValue
+		}
+		tags := append(dd.tags, fmt.Sprintf(stateTagFormat, string(status.ClusterChecksRunner.State)))
+		if err := dd.sendDeploymentMetric(metricValue, clustercheckrunnerName, tags); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// sendStatusEvents forwards events for each component deployment (agent, clusteragent, clustercheck runner)
+// based on the status of DatadogAgentDeployment
+func (dd *MetricsForwarder) sendStatusEvents(status *datadoghqv1alpha1.DatadogAgentDeploymentStatus) error {
+	if status == nil {
+		return errors.New("nil status")
+	}
+
+	// Agent deployment events
+	if status.Agent != nil && !dd.componentEnabled[agentName] {
```

Comment:
I wonder if there's a way to reorganize these if statements for better readability. There's a similar pattern but the types are different.

Reply:
good point, maybe we can add a method (`detectEvent(componentIsNil bool, componentName, componentEventName string)` for example) and we call it this way `dd.detectEvent(status.<component> == nil, <component name>, <component event name>)` 

This way the logic could be shared between the 3 components

Reply:
more simply, we can just pass `componentIsNil` to `sendDeploymentEvent`

Reply:
but adding logic to `delegatedSendDeploymentEvent` will prevent us from testing and asserting that we're calling the datadog api using the correct values 🤔

Reply:
nm, could be solved using this method
```golang
func (dd *MetricsForwarder) getDeploymentEventType(isComponentNil bool, componentName string) *string {
	if !isComponentNil && !dd.componentEnabled[agentName] {
		dd.componentEnabled[agentName] = true
		return &deploymentCreatedEventType
	}
	if isComponentNil && dd.componentEnabled[agentName] {
		dd.componentEnabled[agentName] = false
		return &deploymentDeletedEventType
	}
	return nil
}
```

Reply:
called this way
```golang
	if eventType := dd.getDeploymentEventType(status.Agent == nil, agentName); eventType != nil {
		if err := dd.sendDeploymentEvent(agentEventName, *eventType); err != nil {
			return err
		}
	}
```